chain/chunks/src/lib.rs

```diff
@@ -567,8 +597,7 @@ impl ShardsManager {
         &self,
         parent_hash: &CryptoHash,
         shard_id: ShardId,
-        request_from_archival: bool,
-    ) -> Result<AccountIdOrPeerTrackingShard, near_chain::Error> {
+    ) -> Result<Option<AccountId>, near_chain::Error> {
         let mut block_producers = vec![];
```

Comment:
Not related to this change, but creating `let mut block_producers = vec![];` for the purpose of getting one element seems a bit wasteful.

Reply:
After playing with my branch for a while
```
diff --git i/chain/chunks/src/lib.rs w/chain/chunks/src/lib.rs
index 7b46a51de..dd9f359f6 100644
--- i/chain/chunks/src/lib.rs
+++ w/chain/chunks/src/lib.rs
@@ -8,6 +8,7 @@ use cached::{Cached, SizedCache};
 use chrono::DateTime;
 use log::{debug, error, warn};
 use near_primitives::time::Utc;
+use rand::prelude::IteratorRandom;
 use rand::seq::SliceRandom;
 
 use near_chain::validate::validate_chunk_proofs;
@@ -598,26 +599,28 @@ impl ShardsManager {
         parent_hash: &CryptoHash,
         shard_id: ShardId,
     ) -> Result<Option<AccountId>, near_chain::Error> {
-        let mut block_producers = vec![];
         let epoch_id = self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash).unwrap();
-        for (validator_stake, is_slashed) in
-            self.runtime_adapter.get_epoch_block_producers_ordered(&epoch_id, parent_hash)?
-        {
-            let account_id = validator_stake.take_account_id();
-            if !is_slashed
-                && self.cares_about_shard_this_or_next_epoch(
-                    Some(&account_id),
-                    &parent_hash,
-                    shard_id,
-                    false,
-                )
-                && self.me.as_ref() != Some(&account_id)
-            {
-                block_producers.push(account_id);
-            }
-        }
+        let block_producers =
+            self.runtime_adapter.get_epoch_block_producers_ordered(&epoch_id, parent_hash)?;
+        let block_producers =
+            block_producers.into_iter().filter_map(|(validator_stake, is_slashed)| {
+                let account_id = validator_stake.take_account_id();
+                if !is_slashed
+                    && self.cares_about_shard_this_or_next_epoch(
+                        Some(&account_id),
+                        &parent_hash,
+                        shard_id,
+                        false,
+                    )
+                    && self.me.as_ref() != Some(&account_id)
+                {
+                    Some(account_id)
+                } else {
+                    None
+                }
+            });
 
-        Ok(block_producers.choose(&mut rand::thread_rng()).cloned())
+        Ok(block_producers.choose(&mut rand::thread_rng()).clone())
     }
 
     fn get_tracking_shards(&self, parent_hash: &CryptoHash) -> HashSet<ShardId> {

```

Reply:
Btw, `get_epoch_block_producers_ordered` can be optimized as well, to not return vec and avoid allocation.

Reply:
I'll make another PR for this change

Reply:
https://github.com/near/nearcore/pull/5689